packages/backend/src/gerrit.ts

```diff
@@ -0,0 +1,179 @@
+import fetch from 'cross-fetch';
+import { GerritConfig } from './schemas/v2.js';
+import { AppContext, GitRepository } from './types.js';
+import { createLogger } from './logger.js';
+import path from 'path';
+import { measure, marshalBool } from './utils.js';
+import micromatch from "micromatch";
+
+// TODO: Use gerrit API: https://github.com/gerritkit/client
+// https://gerrit-review.googlesource.com/Documentation/rest-api.html
+interface GerritProjects {
+   [projectName: string]: GerritProjectInfo;
+}
+
+interface GerritProjectInfo {
+   id: string;
+   state?: string;
+   web_links?: GerritWebLink[];
+}
+
+interface GerritWebLink {
+   name: string;
+   url: string;
+}
+
+interface GerritBranch {
+   ref: string;
+   revision: string;
+   web_links?: GerritWebLink[];
+}
+
+// TODO: Implement tags
+// interface GerritTagger {
+//    name: string;
+//    email: string;
+//    date: string;
+//    tz: number;
+// }
+
+// interface GerritTag {
+//    ref: string;
+//    revision: string;
+//    object: string;
+//    message: string;
+//    tagger: GerritTagger;
+// }
+
+const logger = createLogger('Gerrit');
+
+export const getGerritReposFromConfig = async (config: GerritConfig, ctx: AppContext) => {
+   // Example URLs for experimentation:
+   // https://chromium-review.googlesource.com
+   // https://review.opendev.org
+   // https://android-review.googlesource.com
+
+   const url = config.url.endsWith('/') ? config.url : `${config.url}/`;
+
+   const { durationMs, data: projects } = await measure(() =>
+      fetchProjects(url, config)
+   );
+
+   logger.debug(`Fetched ${Object.keys(projects).length} projects in ${durationMs}ms.`);
+
+   let projectNames = Object.keys(projects);
+
+   // If specific repos are specified in config, filter the projects
+   if (config.projects && config.projects.length > 0) {
+      projectNames = projectNames.filter((name) => config.projects!.includes(name));
```

Comment:
nit: instead of getting all projects and then filtering that list down, could we instead specifically fetch each of the `project`s individually using [this api](https://gerrit-review.googlesource.com/Documentation/rest-api-projects.html#:~:text=Projects%20REST%20endpoint.-,Get%20Project,-%27GET%20/projects/)?

Reply:
ack

Reply:
If we do glob patterns, fetching all projects makes sense